Update PyROS Uncertainty Set Validation Methods

jas-yao opened the pull request:
## Fixes: #2724, #3508 

## Summary/Motivation:
This PR provides updates to PyROS uncertainty set validation methods and related tests.
Here, a `validate` method replaces the `is_valid` method (which solves 2N bounding problems to check for set boundedness) in all uncertainty sets, with each set having its own custom `validate` method that efficiently checks set-specific attributes and raises informative exceptions if any issues are found.


## Changes proposed in this PR:
- Update `is_bounded` and `is_nonempty` methods in base `UncertaintySet` class
- Provide a `_solve_feasibility` method in base `UncertaintySet` class
- Replace `is_valid` with `validate` method that runs `is_bounded` and `is_nonempty` in the base `UncertaintySet` class
- Override `validate` in subclass uncertainty sets to check set-specific attributes
- Remove attribute setter checks in uncertainty sets that have been moved to the `validate` method
- Update unit tests for `is_bounded`, `is_nonempty`, `_solve_feasibility`, and `validate` methods

## TODO
- [x] Add FBBT for `_compute_parameter_bounds`
- [x] Add `validate_array` to all `validate` methods
- [x] Split `test_validate` tests into separate tests for each validation check
- [x] Delete `_validate` method for `PolyhedralSet`
- [x] Update error messages to clearly indicate issues specific to uncertainty set setup
- [x] Check uncertainty set documentation looks good and test fix for #3508 
- [ ] Update version number and changelog (after #3581 merged)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

jas-yao commented:
@shermanjasonaf

jas-yao commented:
Thanks for all the feedback and suggestions @jsiirola !
Regarding FBBT, I was able to reproduce the issue with the following short script.

```python
"""Reproducing FBBT error."""
from pyomo.contrib.pyros.uncertainty_sets import PolyhedralSet
from pyomo.contrib.fbbt.fbbt import fbbt

# create a polyhedral uncertainty set that is not bounded
lhs_coefficients_mat = [[1, 2, 3], [4, 5, 6]]
rhs_vec = [1, 3]
pset = PolyhedralSet(lhs_coefficients_mat, rhs_vec)

# create pyomo model and pprint
bounding_model = pset._create_bounding_model()
bounding_model.pprint()

# try to use fbbt and pprint results
fbbt(bounding_model)
bounding_model.pprint()
```

with traceback

```python
Traceback (most recent call last):
  File "<string>", line 17, in __PYTHON_EL_eval
  File "/Users/yletian/Documents/projects/pyomo/pyomo/contrib/pyros/temp_fbbt_error.py", line 15, in <module>
    fbbt(bounding_model)
    ~~~~^^^^^^^^^^^^^^^^
  File "/Users/yletian/Documents/projects/pyomo/pyomo/contrib/fbbt/fbbt.py", line 1543, in fbbt
    _new_var_bounds = _fbbt_block(comp, config)
  File "/Users/yletian/Documents/projects/pyomo/pyomo/contrib/fbbt/fbbt.py", line 1429, in _fbbt_block
    _new_var_bounds = _fbbt_con(c, config)
  File "/Users/yletian/Documents/projects/pyomo/pyomo/contrib/fbbt/fbbt.py", line 1334, in _fbbt_con
    visitorA.walk_expression(con.expr)
    ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/Users/yletian/Documents/projects/pyomo/pyomo/core/expr/visitor.py", line 276, in walk_expression
    result = self._process_node(root, RECURSION_LIMIT)
  File "/Users/yletian/Documents/projects/pyomo/pyomo/core/expr/visitor.py", line 472, in _process_node_bx
    tmp = self.beforeChild(node, child, child_idx)
  File "/Users/yletian/Documents/projects/pyomo/pyomo/contrib/fbbt/fbbt.py", line 1160, in beforeChild
    return _before_child_handlers[child.__class__](self, child)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "/Users/yletian/Documents/projects/pyomo/pyomo/contrib/fbbt/fbbt.py", line 1108, in _register_new_before_child_handler
    if child.is_variable_type():
       ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'numpy.int64' object has no attribute 'is_variable_type'
```

This was fixed with your suggested changes:
```diff
@@ -19,7 +19,7 @@ from pyomo.core.expr.visitor import (
     identify_variables,
     StreamBasedExpressionVisitor,
 )
-from pyomo.core.expr.numvalue import nonpyomo_leaf_types, value
+from pyomo.common.numeric_types import nonpyomo_leaf_types, native_numeric_types, value
 from pyomo.core.expr.numvalue import is_fixed
 import pyomo.contrib.fbbt.interval as interval
 import math
@@ -1105,7 +1105,9 @@ def _before_external_function(visitor, child):
 def _register_new_before_child_handler(visitor, child):
     handlers = _before_child_handlers
     child_type = child.__class__
-    if child.is_variable_type():
+    if child_type in native_numeric_types:
+        handlers[child_type] = _before_constant
+    elif child.is_variable_type():
         handlers[child_type] = _before_var
     elif not child.is_potentially_variable():
         handlers[child_type] = _before_NPV
```

With this fix, I'll take a deeper look into possibly replacing the 2N bounding problems of `is_bounded` with FBBT.

blnicho commented:
@jas-yao is this ready for review or should we wait for the TODO's in the description to be completed?

jas-yao commented:
Hi @blnicho, I will need to finish some of the TODOs before this will be ready for review.

shermanjasonaf commented:
@jas-yao Please delay the version number/changelog update until after #3581 has been merged.

jas-yao commented:
@shermanjasonaf, @blnicho, @jsiirola I have completed the TODOs and think this PR should be ready for review now.

jsiirola reviewed (commented):
Over all, this looks pretty good.  A couple small questions / edits and I think we can be good to go.

shermanjasonaf commented on pyomo/contrib/pyros/tests/test_uncertainty_sets.py:
```diff
@@ -2623,6 +3050,88 @@ def test_compute_parameter_bounds(self):
+        time_with_bounds_provided = end - start
+
+        # when parameter_bounds is not available
+        custom_set.parameter_bounds = None
```
The `parameter_bounds` attribute should be an empty list if no bounds are to be provided:

```suggestion
        custom_set.parameter_bounds = []
```

shermanjasonaf commented on pyomo/contrib/pyros/tests/test_uncertainty_sets.py:
```diff
@@ -2623,6 +3050,88 @@ def test_compute_parameter_bounds(self):
+        self.assertLess(
+            time_with_bounds_provided,
+            time_without_bounds_provided,
+            "Boundedness check with provided parameter_bounds took longer than expected.",
```
Wrap this string to reduce the line length

```suggestion
            "Boundedness check with provided parameter_bounds took longer than "
            "expected.",
```

jas-yao replied:
This time-based test has been removed

jas-yao commented on pyomo/contrib/pyros/uncertainty_sets.py:
```diff
@@ -2519,12 +2837,12 @@ class EllipsoidalSet(UncertaintySet):
+           [0, 0, 3, 0],
+           [0, 0, 0, 4]])
     >>> conf_ellipsoid.scale
     ...9.4877...
```
I ran my doctest with 
```bash
python -m doctest -o=ELLIPSIS
```
which is why they passed. I'm not sure how doctests will normally be run, so it will be good to include these to ensure they pass.

jas-yao commented on pyomo/contrib/pyros/tests/test_uncertainty_sets.py:
```diff
@@ -2623,6 +3050,88 @@ def test_compute_parameter_bounds(self):
+
+        # check with parameter_bounds should always take less time than solving 2N
+        # optimization problems
+        self.assertLess(
```
This test is not required for verifying that the `is_bounded` method is working. I have removed it to avoid dealing with the hassle of system-dependent results.

jas-yao commented on pyomo/contrib/pyros/uncertainty_sets.py:
```diff
@@ -558,23 +567,38 @@ def is_bounded(self, config):
-        param_bounds_arr = np.array(
-            self._compute_parameter_bounds(solver=config.global_solver)
-        )
+        # use parameter bounds if they are available
+        param_bounds_arr = self.parameter_bounds
+        if param_bounds_arr:
+            all_bounds_finite = np.all(np.isfinite(param_bounds_arr))
+        else:
+            # initialize uncertain parameter variables
+            param_bounds_arr = np.array(self._fbbt_parameter_bounds(config))
+            if not all(map(lambda x: all(x), param_bounds_arr)):
+                # solve bounding problems if FBBT cannot find bounds
+                param_bounds_arr = np.array(
+                    self._compute_parameter_bounds(solver=config.global_solver)
+                )
+            all_bounds_finite = np.all(np.isfinite(param_bounds_arr))
```
This may have an undesirable behavior of returning True, even if `self.parameter_bounds` contains bad bounds (e.g inf, nan) because it will continue to try to find better bounds with `_fbbt_paramter_bounds` and `_compute_parameter_bounds`. I think it might be better to keep the cases of `self.parameter_bounds` exists and doesn't exist separate.

How about this?
```python
        # use parameter bounds if they are available
        param_bounds_arr = np.array(self.parameter_bounds)
        if param_bounds_arr.size > 0:
            all_bounds_finite = (
                np.isfinite(param_bounds_arr).all()
            )
        else:
            # use FBBT
            param_bounds_arr = np.array(
                self._fbbt_parameter_bounds(config),
                dtype="float",
            )
            all_bounds_finite = np.isfinite(param_bounds_arr).all()

            if not all_bounds_finite:
                # solve bounding problems
                param_bounds_arr = np.array(
                    self._compute_parameter_bounds(solver=config.global_solver)
                )
                all_bounds_finite = np.isfinite(param_bounds_arr).all()
```

jas-yao commented on pyomo/contrib/pyros/uncertainty_sets.py:
```diff
@@ -558,23 +567,38 @@ def is_bounded(self, config):
+            if not all(map(lambda x: all(x), param_bounds_arr)):
+                # solve bounding problems if FBBT cannot find bounds
+                param_bounds_arr = np.array(
+                    self._compute_parameter_bounds(solver=config.global_solver)
```
I have chosen to go with option 1 in using an index list of 2-tuple of bool for this. Here, I have the check to run/skip optimization problems directly in the for loop using this index list. I have completely removed the inefficiency mentioned by @jsiirola.

I think that doing it this way has not caused any issues with other parts of the PyROS codebase outside of `is_bounded`, since the `_compute_parameter_bounds` return object is the same as before.

jas-yao commented:
@shermanjasonaf @jsiirola I have addressed all comments. 

I would appreciate further feedback on my implementations for `is_bounded` and `_compute_exact_parameter_bounds`. 
I will also need to complete the final TODO to update the version number and changelog.

jsiirola reviewed (approved):
This looks good.  @jas-yao, @shermanjasonaf: do you want to update the version & changelog as part of this PR before we merge it, or just add it to the next PR?

mrmundt commented on pyomo/contrib/pyros/uncertainty_sets.py:
```diff
@@ -583,16 +620,56 @@ def is_bounded(self, config):
+
+        Returns
+        -------
+        : bool
```
The actual varname is missing here.

mrmundt commented on pyomo/contrib/pyros/uncertainty_sets.py:
```diff
@@ -508,6 +511,14 @@ def parameter_bounds(self):
+
+        Returns
+        -------
+        : list of tuple
```
The actual varname is missing here.

mrmundt reviewed (approved):
There are two doc issues (missing variables names), but this is otherwise lovely.

shermanjasonaf commented:
@jas-yao This looks good. You may want to remove the custom implementation of `DiscreteScenarioSet.is_bounded()`, which, though never used in the code (apart from in the discrete uncertainty set tests), unconditionally returns `True`.

shermanjasonaf commented:
> This looks good. @jas-yao, @shermanjasonaf: do you want to update the version & changelog as part of this PR before we merge it, or just add it to the next PR?

@jsiirola Let's leave the version number and changelog updates for #3649.
